lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/TaskHostBase.cs

```diff
@@ -0,0 +1,229 @@
+                }
+            }
+            finally
+            {
+                FinallyBlock();
+                _taskCloseCoordinator.SignalTaskStopped();
+            }
+            Logger.Log(Level.Info, "{0} returned with cancellation token:{1}.", TaskHostName, _cancellationSource.IsCancellationRequested);
+            return null;
+        }
+
+        private static bool IsCommunicationException(Exception e)
+        {
+            if (e is OperationCanceledException || e is IOException || e is TcpClientConnectionException ||
+                e is RemotingException || e is SocketException ||
+                (e is AggregateException && e.InnerException != null && e.InnerException is IOException))
+            {
+                return true;
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// The body of Call method. Subclass must override it. 
+        /// </summary>
+        /// <param name="memento"></param>
+        /// <returns></returns>
+        protected abstract byte[] TaskBody(byte[] memento);
+        
+        /// <summary>
+        /// The code that needs to be executed no matter exception happens or not in Call() method.  
+        /// </summary>
+        protected virtual void FinallyBlock()
+        {            
+        }
+
+        /// <summary>
+        /// Task host name
+        /// </summary>
+        protected abstract string TaskHostName { get; }
+
+        /// <summary>
+        /// Task close handler. Call TaskCloseCoordinator to handle the event.
+        /// </summary>
+        /// <param name="closeEvent"></param>
+        public void OnNext(ICloseEvent closeEvent)
+        {
+            _taskCloseCoordinator.HandleEvent(closeEvent, _cancellationSource);
+        }
+
+        /// <summary>
+        /// Dispose function. Dispose IGroupCommunicationsClient.
+        /// </summary>
+        public void Dispose()
+        {
+            if (Interlocked.Exchange(ref _disposed, 1) == 0)
+            {
+                _groupCommunicationsClient.Dispose();
+            }
+        }
+
+        public void OnError(Exception error)
+        {
+            throw new NotImplementedException();
+        }
+
+        public void OnCompleted()
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <summary>
+        /// Convert the exception into IMRUTaskGroupCommunicationException
+        /// </summary>
+        /// <param name="e"></param>
+        protected void HandleCommunicationException(Exception e)
+        {
+            HandleException(e, new IMRUTaskGroupCommunicationException(TaskManager.TaskGroupCommunicationError, e));
+        }
+
+        /// <summary>
+        /// Convert the exception into IMRUSystemException
+        /// </summary>
+        /// <param name="e"></param>
+        protected void HandleSystemException(Exception e)
+        {
+            HandleException(e, new IMRUSystemException(TaskManager.TaskSystemError, e));
```

Comment:
We should use `IMRUTaskSystemException` here, which corresponds to task non-communication non-app exception. `IMRUSystemException` is "Type of exception thrown when possible bugs are detected in IMRU code." which should be thrown in our logic, not wrap client code.